docs/developer_overview.md

```diff
@@ -0,0 +1,527 @@
+      : temp_storage(PrivateStorage()),
+        linear_tid(RowMajorTid(BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z)) {}
+
+  __device__ __forceinline__ BlockReduce(TempStorage &temp_storage)
+      : temp_storage(temp_storage.Alias()),
+        linear_tid(RowMajorTid(BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z)) {}
+};
+```
+
+Unlike warp-scope facilities, block-scope ones provide a default constructor that allocates memory internally:
+
+```cpp
+__device__ __forceinline__ _TempStorage& PrivateStorage()
+{
+  __shared__ _TempStorage private_storage;
+  return private_storage;
+}
+```
+
+The `__shared__` memory has static semantic, so it's safe to return a reference here.
+
+#### Specialization
+
+Block-scope facilities usually expose algorithm selection to the user. The algorithm is represented by the enumeration part of the API. For the reduction case, `BlockReduceAlgorithm` is provided. Specializations are stored in the `cub/block/specializations` directory. 
+
+#### Temporary storage usage
+
+For block-scope algorithms, it's unsafe to use temporary storage without synchronization:
+
+```cpp
+using BlockReduce = cub::BlockReduce<int, 128> ;
+
+__shared__ BlockReduce::TempStorage temp_storage;
+
+int aggregate_1 = BlockReduce(temp_storage).Sum(thread_data_1);
+// illegal, has to add `__syncthreads` between the two
+int aggregate_2 = BlockReduce(temp_storage).Sum(thread_data_2);
+// illegal, has to add `__syncthreads` between the two
+foo(temp_storage);
+```
+
+## Device-scope
+
+#### Overview
+
+Device-scope functionality is provided by static member functions:
+
+```cpp
+struct DeviceReduce
+{
+
+
+template <typename InputIteratorT, typename OutputIteratorT>
+CUB_RUNTIME_FUNCTION static 
+cudaError_t Sum(
+    void *d_temp_storage, size_t &temp_storage_bytes, InputIteratorT d_in,
+    OutputIteratorT d_out, int num_items, cudaStream_t stream = 0) 
+{
+  using OffsetT = int;
+  using OutputT =
+      cub::detail::non_void_value_t<OutputIteratorT,
+                                    cub::detail::value_t<InputIteratorT>>;
+
+  return DispatchReduce<InputIteratorT, OutputIteratorT, OffsetT,
+                        cub::Sum>::Dispatch(d_temp_storage, temp_storage_bytes,
+                                            d_in, d_out, num_items, cub::Sum(),
+                                            OutputT(), stream);
+}
+
+
+};
+```
+
+Device-scope facilities always return `cudaError_t` and accept `stream` as the last parameter (main stream by default). First two parameters are always `void *d_temp_storage, size_t &temp_storage_bytes`. The algorithm invocation consists of two phases. During the first phase, temporary storage size is calculated and returned in `size_t &temp_storage_bytes`. During the second phase, `temp_storage_bytes` of memory is expected to be allocated and `d_temp_storage` is expected to be the pointer to this memory.
+
+> :warning: Even if the algorithm doesn't need shared memory, we still require one byte of memory to be allocated.
```

Comment:
I've used the general [term](https://en.wikipedia.org/wiki/Shared_memory#:~:text=In%20computer%20science%2C%20shared%20memory,of%20passing%20data%20between%20programs.) here. I agree, that it might be ambiguous in the CUDA context.